proposals/per_policy_webhook_configurations.md

```diff
@@ -0,0 +1,235 @@
+
+Benefits of policy based webhooks:
+1. offers the flexibility for webhook configurations
+2. prevents unexpected webhook failures
+3. reduces the admission review processing duration
+
+# Proposal
+
+There are two dimensions for this feature:
+
+1. offer users the option to configure webooks' attributes including `matchConditions`, `objectSelector` and `namespaceSelector` based on the policy settings
+2. register webhooks based on the policy type, specifically to configure the `scope` attribute of webhooks for namespaced policies
+
+The following snippets list out how policy configurations are transformed to the webhook configurations.
+
+**a. selectors that filters resource objects**
+
+The existing attribute `spec.rules.match.(any/all).resources.selector` can be transformed to the webhooks `objectSelector`.
+```yaml
+spec:
+  rules:
+    - match:
+        all:
+        - resources:
+            selector:
+              matchLabels:
+                foo: bar
+```
+The above rule will be transformed to the following webhook configuration.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+- name: my-webhook.example.com
+  objectSelector:
+    matchLabels:
+      foo: bar
+```
+
+**b. selectors that filters the resources' namespaces**
+The existing attribute `spec.rules.match.(any/all).resources.namespaceSelector` can be transformed to the webhooks `namespaceSelector`.  
+
+```yaml
+spec:
+  rules:
+  - match:
+      all:
+      - resources:
+          namespaceSelector:
+            matchExpressions:
+            - key: foo.com/managed-state
+              operator: In
+              values:
+              - managed
+```
+The above rule will be transformed to the following webhook configuration.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+  - name: my-webhook.example.com
+    namespaceSelector:
+      matchExpressions:
+        - key: foo.com/managed-state
+          operator: In
+          values: ["managed"]
+```
+**c. configure the webhooks' `matchConditions` via CEL expressions (requires Kubernetes 1.27+)**
+The new attribute will be added to Kyverno policy `spec.rules.match.(any/all).resources.matchConditions`.
+```yaml
+spec:
+  rules:
+  - match:
+      all:
+      - resources:
+          matchConditions:
+            matchExpressions:
+            # Match requests made by non-node users.
+            - name: 'exclude-kubelet-requests'
+              expression: '!("system:nodes" in request.userInfo.groups)' 
+```
+The above rule will be transformed to the following webhook configuration.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+  - name: my-webhook.example.com
+    # note: Kubernetes allows up to 64 matchConditions per webhook
+    matchConditions:
+    - name: 'exclude-kubelet-requests'
+      expression: '!("system:nodes" in request.userInfo.groups)' 
```

Comment:
All webhook `matchConditions` must evaluate to true. It isn't OR-ed operation so we will have the same issue as above. 
I believe configuring webhooks `objectSelector`, `namespaceSelector` and `matchConditions` from the policy settings isn't a good idea as we will end up having multiple webhook configurations.